### PR TITLE
Support for Generic Schemes

### LINKED ISSUE #156: implement GenericSchemeProposal

### PR DESCRIPTION
**Pull request by jellegerbrandy**
- proposal.search now understands teh `type` argument
- fixes #156

### CODE REVIEW

File: src/arc.ts [hunk 1]
```diff
@@ -384,7 +400,7 @@ export class Arc {
   }
 
   public setAccount(address: Address) {
-    this.web3.eth.accounts.wallet[0] = address
+    // this.web3.eth.accounts.wallet[0] = address
```
Comment (tibetsprague) [inline src/arc.ts]: remove?

File: src/proposal.ts [hunk 2]
```diff
@@ -94,80 +127,195 @@ export class Proposal implements IStateful<IProposalState> {
...
+          options.descriptionHash = await saveIPFSData()
+          const transaction = schemeRegistrar.methods.proposeToRemoveScheme(
+            options.dao,
+            options.scheme,
+            options.descriptionHash
+          )
+          return transaction
+        }
+        break
+              // function proposeToRemoveScheme(Avatar _avatar, address _scheme, string memory _descriptionHash)
```
Comment (tibetsprague) [inline src/proposal.ts]: remove

File: src/proposal.ts [hunk 3]
```diff
@@ -94,80 +127,195 @@ export class Proposal implements IStateful<IProposalState> {
...
     context: Arc,
     apolloQueryOptions: IApolloQueryOptions = {}
-  ): Observable<Proposal[]> {
+  ): Observable < Proposal[] > {
     let where = ''
+
+    // default options
+    options.type = options.type || IProposalType.ContributionReward
+
+    // constribut the query
```
Comment (tibetsprague) [inline src/proposal.ts]: huh?

File: src/proposal.ts [hunk 4]
```diff
@@ -284,45 +446,79 @@ export class Proposal implements IStateful<IProposalState> {
...
-        reputationReward: new BN(item.contributionReward.reputationReward),
         resolvedAt: item.resolvedAt !== undefined ? Number(item.resolvedAt) : null,
+        schemeRegistrar,
         stage: IProposalStage[item.stage] as any,
         stakesAgainst: new BN(item.stakesAgainst),
         stakesFor: new BN(item.stakesFor),
         thresholdConst: Number(item.thresholdConst),
         title: item.title,
         totalRepWhenExecuted: new BN(item.totalRepWhenExecuted),
+        type: IProposalType.ContributionReward,
```
Comment (tibetsprague) [inline src/proposal.ts]: shouldn't this be the actual type?

File: src/proposal.ts (hunk 3 above)
Comment (jellegerbrandy) [reply src/proposal.ts]: :-)

File: src/queue.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+export interface IQueue {
+  id: string
+  dao: Address
+  name?: string
+}
+
+export interface IQueueState {
+  id: string
+  // scheme: IScheme
```
Comment (tibetsprague) [inline src/queue.ts]: why no scheme?

File: src/queue.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
+import BN = require('bn.js')
+import gql from 'graphql-tag'
+import { Observable } from 'rxjs'
+import { Arc, IApolloQueryOptions } from './arc'
+import { Address } from './types'
+import { realMathToNumber } from './utils'
+
+export interface IQueue {
```
Comment (tibetsprague) [inline src/queue.ts]: this doesnt seem to be used

File: src/queue.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      )
+    }
+
+    return context._getObservableList(query, itemMap, apolloQueryOptions) as Observable<Queue[]>
+  }
+  public id: Address
+  public dao: Address
+  public name: string
+
+  constructor(id: Address, dao: Address, name: string, public context: Arc) {
```
Comment (tibetsprague) [inline src/queue.ts]: I think its always nice to have the constructor at the top of the class, and the instance variables too

File: src/scheme.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        item.name,
+        item.paramsHash,
+        context
+      )
+    }
+
+    return context._getObservableList(query, itemMap) as Observable<Scheme[]>
+  }
+
+  constructor(
```
Comment (tibetsprague) [inline src/scheme.ts]: I guess the bottom is ok too, consistency at least

File: src/queue.ts (hunk 5 above)
Comment (jellegerbrandy) [reply src/queue.ts]: because the scheme properties are now part of the QueueState...

File: src/queue.ts (hunk 7 above)
Comment (jellegerbrandy) [reply src/queue.ts]: my linter does this automatically, and what it does is:

- first static methods
-then instance variables
- then the constructor
- then the normal methods

for me this is fine :-)

**Pull request merged**